examples/matrix_multiplication.masm

```diff
@@ -0,0 +1,727 @@
+            # [H, H, H, addr, c, words, r, ...]
+            swap movdnw.3
+
+            # [H, H, H, addr, c, words, r, ...]
+            adv_pipe
+
+            # INNER: [c, addr, words, r, H, H, H, ...]
+            movupw.3 swap
+
+            # every call to `adv_pipe` copies *two* words
+            sub.2
+
+            # while(c>0)
+            dup exec.is_not_zero_or_negative_one
+        end
+
+        # remove the padding word at the end of the row
+        eq.18446744069414584320 if.true
+            sub.1
+        end
+
+        # OUTER: [r, words, addr, H, H, H, ...]
+        movdn.2 swap
+
+        # while(r>0)
+        sub.1 dup neq.0
+    end
+
+    # [H, H, H, ...]
+    drop drop drop
+end
+
+# matrix_column_major_copy(destination: out address)
+#
+# Copies a matrix from the advice tape to memory. The matrix data must be
+# carefully encoded into the advice tape. The first eigth elements must be:
+#
+#   [0, 0, col, row, 0, 0, 0, 0]
+#
+# Then the matrix is encoded column-wise, from left to right. Each column must
+# be first padded with the element 0 until its size is a multiple of 8. These
+# elements are then handled in groups of 4, in each group their elements be in
+# reverse order. As an example, the following matrix:
+#
+#   [8, 3, 1]
+#   [9, 6, 7]
+#
+# First has its columns padded to be a multiple of 8:
+#
+#   [8, 3, 0, 0, 0, 0, 0, 0]
+#   [9, 6, 0, 0, 0, 0, 0, 0]
+#   [1, 7, 0, 0, 0, 0, 0, 0]
+#
+# Then reverse the elements order for each 4-element group:
+#
+#   [0, 0, 3, 8, 0, 0, 0, 0]
+#   [0, 0, 6, 9, 0, 0, 0, 0]
+#   [0, 0, 7, 1, 0, 0, 0, 0]
+#
+# Resulting in the following encoding:
+#
+# [
+#     0, 0, 3, 2, 0, 0, 0, 0,  # matrix size
+#     0, 0, 3, 8, 0, 0, 0, 0,  # first column
+#     0, 0, 6, 9, 0, 0, 0, 0,  # second column
+#     0, 0, 7, 1, 0, 0, 0, 0   # third column
+# ]
+#
+# The data is read into memory preserving padding *elements*, but removing
+# padding *words*. So the data above will be store as:
+#
+#     addr+0: [2, 3, 0, 0]  # matrix size
+#     addr+1: [8, 3, 0, 0]  # first column
+#     addr+2: [9, 6, 0, 0]  # second column
+#     addr+3: [1, 7, 0, 0]  # third column
+#
+# Notes:
+# - This procedure *must* be called with `exec` to copy the data to memory.
+# - The data packingig done to take advantage of `adv_pipe`, a very efficient
+# primitive to load and verify data from tape to memory. Tightly packing the
+# data into the tape would likely be two orders of magnitude slower.
+#
+proc.matrix_column_major_copy.0
+    # input: [addr, ...]
+    # output: [H, H, H, ...]
+
+    # [H, H, H, addr, 0, 0, 0, ...]
+    push.0 push.0 push.0 movup.3 padw padw padw
+
+    adv_pipe
+
+    # [addr, 0, 0, 0, H, H, H, ...]
+    movupw.3
+
+    # remove the padding word after the [row, column]
+    sub.1
+
+    # [addr, 0, 0, 0, addr, H, H, H, ...]
+    dup movdn.4
+
+    # [row, col, 0, 0, addr, H, H, H, ...]
+    dup sub.1 mem_loadw
+
+    # [row, col, addr, H, H, H, ...]
+    movup.2 drop movup.2 drop
```

Comment:
nit: since this is a duplication of the beginning of the `matrix_row_major_copy`, I probably would have made a helper procedure for the setup. It would be easier to see where these procedures diverge. It's up to you though